src/scroll-seek-controls.js

```diff
@@ -0,0 +1,165 @@
+  let observer = null;
+  let video = null;
+  let onScroll = null;
+  let updateProgressUI = null;
+  let hideTimeout = null;
+
+  const DEBUG = false; // Set to true for verbose logs
+
+  function log(...args) {
+    if (DEBUG) console.log('[ScrollSeek]', ...args);
+  }
+
+  function getDynamicScrollStep() {
+    const duration = video?.duration || 600;
+    const percent = 0.01;
+    return Math.min(15, Math.max(3, duration * percent));
+  }
+
+  function isCarouselVisible() {
+    const visibleContainers = [
+      ...document.querySelectorAll('.ytVirtualListContainer')
+    ].filter((el) => {
+      const style = window.getComputedStyle(el);
+      const rect = el.getBoundingClientRect();
+      return (
+        style.display !== 'none' &&
+        rect.width > 0 &&
+        rect.height > 0 &&
+        style.visibility !== 'hidden'
+      );
+    });
+    return visibleContainers.length > 0;
+  }
+
+  function enableScrollSeek() {
+    if (isInitialized) return;
+
+    const container = document.querySelector('ytlr-progress-bar');
+    const playhead = container?.querySelector('ytlr-playhead');
+    const playedBar = container?.querySelector('.ytLrProgressBarPlayed');
+    const slider = container?.querySelector('#slider');
+    video = document.querySelector('video');
+
+    if (!container || !video) {
+      log('Missing elements, setting up observer...');
+      if (!observer) {
+        observer = new MutationObserver(() => {
+          if (!isInitialized && document.querySelector('video')) {
+            enableScrollSeek();
+          }
+        });
+        observer.observe(document.body, { childList: true, subtree: true });
+      }
+      return;
+    }
+
+    function isProgressVisible() {
+      const style = playhead && window.getComputedStyle(playhead);
+      return playhead && style?.opacity !== '0' && style?.display !== 'none';
+    }
+
+    updateProgressUI = function () {
+      if (!video || !isFinite(video.currentTime) || !isFinite(video.duration))
+        return;
+
+      const percent = video.currentTime / video.duration;
+      const barWidth = container.offsetWidth || 1000;
+      const px = barWidth * percent;
+
+      if (playhead) {
+        playhead.style.transform = `translateX(${px}px)`;
+        playhead.style.opacity = '1';
+      }
+      if (playedBar) {
+        playedBar.style.transform = `translateX(0rem) scaleX(${percent})`;
+        playedBar.style.opacity = '1';
+      }
+      if (slider) {
+        slider.style.opacity = '1';
+      }
+
+      clearTimeout(hideTimeout);
+      hideTimeout = setTimeout(() => {
+        if (playhead) playhead.style.opacity = '0';
+        if (playedBar) playedBar.style.opacity = '0';
+        if (slider) slider.style.opacity = '0';
+      }, 2000);
+    };
+
+    onScroll = function (event) {
+      if (!video || !isFinite(video.duration)) return;
+
+      const carouselVisible = isCarouselVisible();
+      const progressVisible = isProgressVisible();
```

Comment:
Checking whether the progress bar is focused on every scroll event is expensive. Instead, dynamically attach the scroll listener based your condition instead. Additionally, a better way to check if the progress bar is focused, is to watch the `ytlr-progress-bar` element for the `zylon-focus` CSS class. Same with checking if the "carousel" is visible. Using CSS classes instead of computed styles is more efficient in general since checking the former doesn't cause a re-flow.